### ISSUE TITLE
nn_recv blocks forever when called inside fork'd subprocess

### ISSUE THREAD

**Issue by grodtron** [nanomsg/nanomsg#779]
My issue is described in detail at nanomsg/nnpy#18 . It's occuring when using the nnpy nanomsg bindings, but it was suggested by @djc that the root cause is in nanomsg itself, and after a little testing, this seems to be the case.

Essentially what is happening (in pseudo code) is:

```java
sock1 = nn_socket(REQ);
sock1.send(request);
response = sock1.recv;

fork();

if (this_is_child_process()) {
   sock2 = nn_socket(SUB);
   sock2.subscribe("foobar");
   sock2.recv(); // blocks forever.
}
```

The subscribe socket is created in the forked child process, but when calling recv, blocks forever. I know that the problem is not with the publishing socket, because if I remove just the fork and perform the same logic in a single process and a single thread messages are received.

Below are two files `subscriber.c` and `publisher.py` that demonstrate the issue.

1. compiler subcriber.c : `gcc subscriber.c -lnanomsg -o subscriber`
2. run subscriber in one terminal `./subscriber`
3. run publisher.py in a second terminal `./publisher.py`

By simply removing the call to fork (change line 36 from `int pid = fork();` to `int pid = 0;`), you can see the correct behaviour.

**subscriber.c**
```c
#include <nanomsg/nn.h>
#include <nanomsg/pubsub.h>
#include <nanomsg/reqrep.h>

#include <string.h>
#include <assert.h>
#include <unistd.h>
#include <stdio.h>

#define LOG(msg, ...) printf("%s:%d(%d) " msg "\n", __FILE__, __LINE__, getpid())

int main(int argc, const char *argv[])
{
  // Create REQ socket
  LOG("Creating REQ socket");
  int req = nn_socket (AF_SP, NN_REQ);
  assert(req >= 0);
  assert(nn_bind(req, "ipc:///tmp/nn-test-reqrep"));

  // Request the publisher to start publishing
  LOG("Sending start request");
  const char * msg  = "start";
  const int msg_len = strlen(msg);
  assert(msg_len == nn_send(req, msg, msg_len, 0));

  // Await the response
  LOG("Waiting for start response");
  void *buf = NULL;
  const char * response = "started";
  assert(strlen(response) == nn_recv(req, &buf, NN_MSG, 0));
  assert(strncmp(response, buf, strlen(response)) == 0);
  nn_freemsg(buf);

  LOG("Forking");
  // Fork a child process
  int pid = fork();

  // Inside the child, create a subscribe socket and listen to the
  // publisher
  if (pid == 0) {
    LOG("Creating SUB socket");
    int sub = nn_socket (AF_SP, NN_SUB);
    assert(sub >= 0);
    int nbytes;
    void *buf = NULL;

    assert(0 == nn_setsockopt(sub, NN_SUB, NN_SUB_SUBSCRIBE, "topic:", 6));
    assert(nn_connect(sub, "ipc:///tmp/nn-test-pubsub"));

    while (1) {
      LOG("Waiting for subscription");
      nbytes = nn_recv (sub, &buf, NN_MSG, 0);
      printf("Received subscription message %s\n", buf);
      nn_freemsg (buf);
    }
  } else {
    while (1) {
      LOG("Sleep loop");
      sleep(1);
    }
  }

  return 0;
}
```

**publisher.py**
```py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import nnpy
import time

print "c: Creating REP socket"
cmd_socket = nnpy.Socket(nnpy.AF_SP, nnpy.REP)
cmd_socket.connect("ipc:///tmp/nn-test-reqrep")
print "c: Created REP socket"

print "c: waiting for start command"
assert 'start' == cmd_socket.recv()
print "c: got start command"

pub_socket = nnpy.Socket(nnpy.AF_SP, nnpy.PUB)
pub_socket.bind("ipc:///tmp/nn-test-pubsub")
print "c: created publisher socket"

cmd_socket.send("started")
print "c: sent response"

for n in range(10):
    print "c: sending update!"
    pub_socket.send("topic:It is now " + str(time.time()) + " o'clock!")
    time.sleep(0.25)
```

**Comment by grodtron** [comment]
After a bit more research, this seems to be the same underlying issue already discussed in #582 , but use in Python `multiprocessing.Process`s is another real life use case where its necessary to use nanomsg after `fork()`

**Comment by fertesta** [comment]
See issue #653 as well.

**Comment by gdamore** [comment]
This is really another item of #582  nanomsg is not fork-friendly at present.

**Comment by gdamore** [comment]
(Closed this as we are tracking under #582)
